vault/external_tests/audit/audit_test.go

```diff
@@ -0,0 +1,194 @@
+	devices, err := client.Sys().ListAudit()
+	require.NoError(t, err)
+	require.Len(t, devices, 1)
+
+	// Request 3
+	err = client.Sys().EnableAuthWithOptions("userpass", &api.EnableAuthOptions{
+		Type: "userpass",
+	})
+	require.NoError(t, err)
+
+	username := "jdoe"
+	password := "abc123"
+
+	// Request 4
+	_, err = client.Logical().Write(fmt.Sprintf("auth/userpass/users/%s", username), map[string]interface{}{
+		"password": password,
+	})
+	require.NoError(t, err)
+
+	authInput, err := userpass.NewUserpassAuth(username, &userpass.Password{FromString: password})
+	require.NoError(t, err)
+
+	newClient, err := client.Clone()
+	require.NoError(t, err)
+
+	correlationID := "correlation-id-foo"
+	newClient.AddHeader("x-correlation-id", correlationID)
+
+	// Request 5
+	authOutput, err := newClient.Auth().Login(context.Background(), authInput)
+	require.NoError(t, err)
+
+	// Request 6
+	hashedPassword, err := client.Sys().AuditHash(devicePath, password)
+	require.NoError(t, err)
+
+	// Request 7
+	hashedClientToken, err := client.Sys().AuditHash(devicePath, authOutput.Auth.ClientToken)
+	require.NoError(t, err)
+
+	// Request 8
+	hashedAccessor, err := client.Sys().AuditHash(devicePath, authOutput.Auth.Accessor)
+	require.NoError(t, err)
+
+	// Request 9
+	wrapResp, err := client.Logical().Write("sys/wrapping/wrap", map[string]interface{}{
+		"foo": "bar",
+	})
+	require.NoError(t, err)
+
+	// Request 10
+	hashedBar, err := client.Sys().AuditHash(devicePath, "bar")
+	require.NoError(t, err)
+
+	hashedWrapAccessor, err := client.Sys().AuditHash(devicePath, wrapResp.WrapInfo.Accessor)
+	require.NoError(t, err)
+
+	hashedWrapToken, err := client.Sys().AuditHash(devicePath, wrapResp.WrapInfo.Token)
+	require.NoError(t, err)
+
+	hashedCorrelationID, err := client.Sys().AuditHash(devicePath, correlationID)
+	require.NoError(t, err)
+
+	entries := make([]map[string]interface{}, 0)
+	scanner := bufio.NewScanner(logFile)
+
+	for scanner.Scan() {
+		entry := make(map[string]interface{})
+
+		err := json.Unmarshal(scanner.Bytes(), &entry)
+		require.NoError(t, err)
+
+		entries = append(entries, entry)
+	}
+
+	// This count includes the initial test probe upon creation of the audit device
+	require.Equal(t, 26, len(entries))
+
+	loginReqEntry := entries[8]
+	loginRespEntry := entries[9]
+
+	loginRequestFromReq := loginReqEntry["request"].(map[string]interface{})
+	loginRequestDataFromReq := loginRequestFromReq["data"].(map[string]interface{})
+	loginHeadersFromReq := loginRequestFromReq["headers"].(map[string]interface{})
+
+	loginRequestFromResp := loginRespEntry["request"].(map[string]interface{})
+	loginRequestDataFromResp := loginRequestFromResp["data"].(map[string]interface{})
+	loginHeadersFromResp := loginRequestFromResp["headers"].(map[string]interface{})
+
+	loginAuth := loginRespEntry["auth"].(map[string]interface{})
+
+	require.True(t, strings.HasPrefix(loginRequestDataFromReq["password"].(string), "hmac-sha256:"))
```

Comment:
❓ would it be 'safer' to pull the `"hmac-sha256:"` prefix into a const string somewhere in here and just reference that in all the assertions?

Reply:
I went back and forth on it. Your feedback has made me decide yes. I'll add that. Thanks!